### PR TITLE
[HttpKernel] Use Accept-Language header even if there are no enabled locales

### LINKED ISSUE #47355: framework.enabled_locales does not behave as documented
### Symfony version(s) affected

5.4.11

### Description

https://symfony.com/doc/5.4/reference/configuration/framework.html#enabled-locales states that "empty array = enable all locales", but if I set an empty array then [set_locale_from_accept_language](https://symfony.com/doc/5.4/reference/configuration/framework.html#set-locale-from-accept-language) will not set the locale. I have to fill `enabled_locales` for it to work.

Here is the code doing this: https://github.com/symfony/symfony/blob/e1581a0937c7e0c8d8420808ebc60418a63892c4/src/Symfony/Component/HttpKernel/EventListener/LocaleListener.php#L71 which was added in #43108 (`$this->enabledLocales` is checked to be non-empty)

@chalasr is this logic on purpose? Or does the documentation need adjustment that empty means no locale is enabled?

### How to reproduce

Run the following I guess: https://github.com/symfony/symfony/blob/123b1651c4a7e219ba59074441badfac65525efe/src/Symfony/Component/HttpKernel/Tests/EventListener/LocaleListenerTest.php#L172

which tests that no locale is set if `enabled_locales` isn't set

### Possible Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by MatTheCat**
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #47355
| License       | MIT
| Doc PR        | N/A

Don’t really know how I should consider this PR :thinking: 

It would affect people setting `set_locale_from_accept_language` to `true` with no `enabled_locales` when the request has a preferred language:

- before: set the default locale as request locale
- after: set the preferred language as request locale

### CODE REVIEW

**Comment by chalasr** [comment]
Can you add a test case that breaks before your patch?

**Comment by MatTheCat** [comment]
@chalasr done.

While updating `testRequestAttributeLocaleNotOverridenFromAcceptLanguageHeader` I stumbled into a weird behavior which I think was caused by the `Accept-Language` header set the wrong way, so I updated tests doing this.

**Comment by fabpot** [comment]
@MatTheCat Can you create a PR on 4.4 for the tests fix?

**Comment by MatTheCat** [comment]
@fabpot see https://github.com/symfony/symfony/pull/47638 (concerned tests were added by https://github.com/symfony/symfony/pull/43108 on 5.4). Should I copy its changes on this PR? Else tests will fail until it is merged and rebased :thinking:

**Comment by chalasr** [comment]
YOu sh

**Comment by chalasr** [comment]
Looks like because #47638's description included "Fix #47377(some comment)", github closed this PR automatically when merging #47638.

> Should I copy its changes on this PR? Else tests will fail until it is merged and rebased 🤔

Rebase unlocked, #47638 has been merged up to 6.2.

**Comment by MatTheCat** [comment]
Damn okay sorry :sweat_smile: 

PR rebased!

**Comment by chalasr** [comment]
No worries, I'd say it should be improved on Github :)
Can you please add a small entry in the framework-bundle CHANGELOG?

**Comment by MatTheCat** [comment]
> Set locale from request rather that default locale if "framework.enabled_locales" is empty and "framework.set_locale_from_accept_language" is true

WDYT? Don’t really know how to make it shorter.

**Comment by chalasr** [comment]
Works for me!

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]:
```suggestion
 * Set locale from request rather than default locale if `framework.enabled_locales` is empty and `framework.set_locale_from_accept_language` is true
```

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md
Comment (MatTheCat) [reply src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: I copied this style from the line above; should I also update it?

Comment (chalasr) [reply src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: Please do if you can yes 👍

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @MatTheCat.

**Comment by nicolas-grekas** [comment]
I'm a bit late here but this means we are going to always send the Vary: Accept-Language header, isn't it? I'm not sure this is legit.

**Comment by MatTheCat** [comment]
@nicolas-grekas only if `set_locale_from_accept_language` is true and the `Accept-Language` header contains some locales.

**Comment by nicolas-grekas** [comment]
Oh, thanks. Two other issues: why on 6.2 and not on 6.1? And more importantly: the vary header should also be sent when $preferredLanguage is falsy. The code should look like this:
```php
        } elseif ($this->useAcceptLanguageHeader) {
            if ($preferredLanguage = $request->getPreferredLanguage($this->enabledLocales)) {
                $request->setLocale($preferredLanguage);
            }
            $request->attributes->set('_vary_by_language', true);
        }
``` 

Can you please send a PR on 6.1 if you agree?

**Comment by MatTheCat** [comment]
I’m not sure about you two questions :disappointed: so I’ll wait for the approval of other members before submitting another PR.

**Comment by chalasr** [comment]
Nicolas is right, please go ahead if you want/can.

**Pull request merged**
